airflow/contrib/hooks/pagerduty_hook.py

```diff
@@ -0,0 +1,129 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""Hook for creating Pagerduty incidents."""
+from typing import Any, Dict, List, Optional
+
+import pypd
+
+from airflow.exceptions import AirflowException
+from airflow.hooks.base_hook import BaseHook
+
+
+class PagerdutyHook(BaseHook):
+    """
+    Takes both Pagerduty API token directly and connection that has Pagerduty API token.
+
+    If both supplied, Pagerduty API token will be used.
+
+    :param token: Pagerduty API token
+    :param pagerduty_conn_id: connection that has Pagerduty API token in the password field
+    """
+
+    def __init__(self, token: Optional[str] = None, pagerduty_conn_id: Optional[str] = None) -> None:
+        self.routing_key = None
+
+        if pagerduty_conn_id is not None:
+            conn = self.get_connection(pagerduty_conn_id)
+
+            self.token = getattr(conn, 'password', None)
+
+            routing_key = conn.extra_dejson.get("routing_key")
+            if routing_key:
+                self.routing_key = routing_key
+
+        if token is not None:  # token takes higher priority
+            self.token = token
+
+        if self.token is None:
+            raise AirflowException(
+                'Cannot get token: No valid api token nor pagerduty_conn_id supplied.')
+
```

Comment:
We can simplify the logic:

```python
        if token is not None:  # token takes higher priority
            self.token = token
        else:
            raise AirflowException(
                'Cannot get token: No valid api token nor pagerduty_conn_id supplied.')
```

Reply:
Making this change will alter the actual logic. Because it will throw an error when token is `None` when there is a valid password set in connection. We want token to override password from connection, but it's not required.

This case has been covered in test cases as well.